File: src/ardldml/diagnostics.py
"""
The trend-absorption diagnostic.

This is the most important thing in the package to actually run, and the
easiest to skip.

The problem it addresses
------------------------
Residualisation is valid only when the controls remove *nuisance* stochastic
trends, not the equilibrium relation being tested. If a control is itself part
of the equilibrium system, partialling it out absorbs the very trend the test
is meant to detect, driving the effective integrated count toward zero and
destroying the relation rather than cleaning it. The estimand changes: a
non-rejection then reflects over-absorption, not the absence of a long-run
relationship.

That requirement -- the paper's Assumption 5, that the nuisance space "does not
span the cointegrating relation of interest" -- cannot be verified directly.
The diagnostic is the practical substitute.

Definition 2
------------
Let :math:`p_{ad}` and :math:`p_{ols}` be the bootstrap p-values under the
adaptive and the unpenalised :math:`m_Z` projection, and :math:`p_{full}` and
:math:`p_{red}` the bootstrap p-values under the full nuisance set and under a
reduced set omitting the controls most likely to be cointegrated with the
tested relation. The diagnostic is the pair of gaps

.. math::
    \\Delta_m = p_{ols} - p_{ad}, \\qquad \\Delta_W = p_{full} - p_{red}

read together with the stability of the long-run coefficient across the four
fits.

How to read it (Remark 9)
-------------------------
* **A large positive** :math:`\\Delta_W` -- a verdict that rejects under the
  reduced set but not under the full set -- together with a long-run
  coefficient that is more stable and more sharply estimated under the reduced
  or adaptive fit, is evidence that the nuisance space is absorbing part of the
  tested relation. Assumption 5 is locally violated and **the reduced-set
  verdict is the more credible one**.
* **Concordant verdicts** across all four fits indicate the conclusion is not
  an artefact of over-absorption.
* The adaptive projection is the default, because vanilla :math:`\\ell_1`
  over-selects integrated regressors and thereby induces spurious absorption.
  The unpenalised projection is retained only to form :math:`\\Delta_m`.

This is a hypothesis-generating device, not a formal test. It has no size and
no power; it tells you where to look.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional, Sequence

import numpy as np
import pandas as pd

from .statistic import DMLBounds

__all__ = ["TrendAbsorption", "trend_absorption", "penalty_sensitivity"]


@dataclass
class TrendAbsorption:
    """
    Result of the four-fit trend-absorption contrast.

    Attributes
    ----------
    fits : dict
        The four fitted :class:`~ardldml.statistic.DMLBoundsResults`, keyed
        ``"full_adaptive"``, ``"full_ols"``, ``"reduced_adaptive"``,
        ``"reduced_ols"``.
    dropped : list of str
        Controls omitted from the reduced set.
    delta_m, delta_W : float
        The two gaps of Definition 2.
    """

    fits: Dict[str, object]
    dropped: List[str]
    delta_m: float
    delta_W: float

    def to_frame(self) -> pd.DataFrame:
        """One row per fit: statistic, critical value, p-value, theta."""
        rows = []
        for key, r in self.fits.items():
            controls, proj = key.split("_")
            rows.append(
                {
                    "controls": controls,
                    "m_Z projection": proj,
                    "n_controls": r.W.shape[1],
                    "F": r.stat,
                    "boot_cv95": r.critical_value,
                    "boot_p": r.pvalue,
                    "verdict": r.decision(),
                    "alpha": r.alpha,
                    "theta": r.theta,
                    "theta_se": r.theta_se,
                }
            )
        return pd.DataFrame(rows)

    @property
    def theta_spread(self) -> float:
        """Range of the long-run coefficient across the four fits."""
        vals = [r.theta for r in self.fits.values() if np.isfinite(r.theta)]
        return float(max(vals) - min(vals)) if vals else float("nan")

    def verdict(self, level: float = 0.05) -> str:
        """
        Plain-language reading, following Remark 9.

        Deliberately conservative: it flags a *possible* violation rather than
        asserting one, because the contrast cannot separate over-absorption
        from ordinary specification sensitivity.
        """
        full = self.fits["full_adaptive"]
        red = self.fits["reduced_adaptive"]
        if full.pvalue is None or red.pvalue is None:
            return "no bootstrap run"

        rejects_reduced = red.pvalue < level
        rejects_full = full.pvalue < level

        if rejects_reduced and not rejects_full and self.delta_W > 0:
            sharper = red.theta_se < full.theta_se
            msg = (
                f"Possible over-absorption. The reduced set rejects "
                f"(p={red.pvalue:.3f}) where the full set does not "
                f"(p={full.pvalue:.3f}); Delta_W = {self.delta_W:+.3f}."
            )
            if sharper:
                msg += (
                    f" The long-run coefficient is also more sharply estimated under the "
                    f"reduced set (se {red.theta_se:.3f} vs {full.theta_se:.3f}), which "
                    f"strengthens the reading. The reduced-set verdict is the more credible."
                )
            else:
                msg += (
                    " The long-run coefficient is not more sharply estimated under the "
                    "reduced set, so ordinary specification sensitivity cannot be ruled out."
                )
            return msg
        if rejects_reduced == rejects_full:
            return (
                f"Concordant verdicts across control sets "
                f"(full p={full.pvalue:.3f}, reduced p={red.pvalue:.3f}). "
                f"The conclusion does not appear to be an artefact of over-absorption."
            )
        return (
            f"Discordant, but not in the over-absorption direction "
            f"(full p={full.pvalue:.3f}, reduced p={red.pvalue:.3f}). "
            f"Treat both verdicts as fragile."
        )

    def summary(self, level: float = 0.05) -> str:
        lines = [
            "Trend-absorption diagnostic (Definition 2)",
            f"dropped from reduced set: {', '.join(self.dropped) or '(none)'}",
            "",
            self.to_frame().to_string(
                index=False, float_format=lambda v: f"{v:.4f}"
            ),
            "",
            f"Delta_m = p_ols  - p_ad  = {self.delta_m:+.4f}",
            f"Delta_W = p_full - p_red = {self.delta_W:+.4f}",
            f"theta spread across fits = {self.theta_spread:.4f}",
            "",
            self.verdict(level),
        ]
        return "\n".join(lines)

    def __repr__(self) -> str:  # pragma: no cover - cosmetic
        return f"<TrendAbsorption delta_W={self.delta_W:+.3f} delta_m={self.delta_m:+.3f}>"


def trend_absorption(
    y: pd.Series,
    d: pd.Series,
    W: pd.DataFrame,
    drop: Sequence[str],
    B: int = 999,
    seed: Optional[int] = None,
    progress: bool = False,
    **spec_kwargs,
) -> TrendAbsorption:
    """
    Run the four-fit contrast of Definition 2.

    Parameters
    ----------
    y, d, W : pandas.Series, pandas.Series, pandas.DataFrame
        Outcome, focal regressor and controls, all in levels.
    drop : sequence of str
        Controls to omit from the reduced set. Choose these on *economic*
        grounds: the ones most likely to be cointegrated with the tested
        relation itself. In the paper's pass-through application these are
        money and the oil price; in the consumption-income appendix they are
        the nominal aggregates M2 and CPI.
    B : int
        Bootstrap replications per fit. Four bootstraps are run, so the total
        cost is roughly ``4 * B`` statistic evaluations.
    seed : int, optional
        Base seed. Each fit is offset so the four bootstraps are independent.
    progress : bool
        Print which fit is running.
    **spec_kwargs
        Passed to :class:`~ardldml.statistic.DMLBoundsSpec`. ``adaptive`` is
        overridden per fit.

    Returns
    -------
    TrendAbsorption
    """
    drop = [c for c in drop if c in W.columns]
    W_red = W.drop(columns=drop)

    integrated = spec_kwargs.pop("integrated", None)
    integrated_red = None
    if integrated is not None:
        integrated = list(integrated)
        integrated_red = [c for c in integrated if c not in drop]

    plan = [
        ("full_adaptive", W, True, integrated, 0),
        ("full_ols", W, False, integrated, 1),
        ("reduced_adaptive", W_red, True, integrated_red, 2),
        ("reduced_ols", W_red, False, integrated_red, 3),
    ]

    fits: Dict[str, object] = {}
    for key, Wk, adaptive, integ, offset in plan:
        if progress:
            print(f"  fitting {key} ({Wk.shape[1]} controls)...", flush=True)
        res = (
            DMLBounds(y, d, Wk, adaptive=adaptive, integrated=integ, **spec_kwargs)
            .fit()
            .bootstrap(B=B, seed=None if seed is None else seed + 10_000 * offset)
        )
        fits[key] = res

    delta_m = float(fits["full_ols"].pvalue - fits["full_adaptive"].pvalue)
    delta_W = float(fits["full_adaptive"].pvalue - fits["reduced_adaptive"].pvalue)
    return TrendAbsorption(fits=fits, dropped=list(drop), delta_m=delta_m, delta_W=delta_W)


def penalty_sensitivity(
    y: pd.Series,
    d: pd.Series,
    W: pd.DataFrame,
    c_grid: Sequence[float] = (1.1, 0.75, 0.5, 0.25),
    lags_grid: Sequence[int] = (4,),
    projections: Sequence[str] = ("adaptive", "ols"),
    B: Optional[int] = None,
    seed: Optional[int] = None,
    **spec_kwargs,
) -> pd.DataFrame:
    """
    Sweep the penalty, the lag order and the ``m_Z`` projection.

    The paper's own robustness tables vary all three, and its Table 14 shows a
    verdict rejecting at one penalty and not at another. Reporting a single
    cell from this grid is specification search; reporting the grid is the
    method.

    The column to watch is ``n_selected_Z``, the number of control **levels**
    the ``m_Z`` projection retained. It is the empirical counterpart of the
    effective integrated count:

    * ``0`` means nothing was absorbed, so the test sits at the classical
      ``k-tilde = k`` corner and orthogonalisation did nothing;
    * a large value means heavy absorption, and the long-run coefficient
      should be checked for instability.

    A long-run coefficient that changes sign across this grid is a warning that
    the conditioning set, not the data, is driving the answer.

    Parameters
    ----------
    c_grid : sequence of float
        Constants in the plug-in penalty. ``1.1`` is the paper's default; lower
        values select more aggressively.
    lags_grid : sequence of int
        Short-run lag orders to try.
    projections : sequence of str
        Any of ``"adaptive"``, ``"plain"``, ``"ols"``.
    B : int, optional
        Bootstrap draws per cell. If ``None``, no bootstrap is run and only the
        statistic and coefficients are reported, which is much faster.

    Returns
    -------
    pandas.DataFrame
        One row per cell.
    """
    rows = []
    for lags in lags_grid:
        for proj in projections:
            for c in c_grid:
                kw = dict(spec_kwargs)
                kw["lags"] = lags
                if proj == "ols":
                    kw["penalised"] = False
                else:
                    kw["penalised"] = True
                    kw["adaptive"] = proj == "adaptive"
                    kw["c"] = c
                try:
                    res = DMLBounds(y, d, W, **kw).fit()
                    if B:
                        res = res.bootstrap(B=B, seed=seed)
                    rows.append(
                        {
                            "lags": lags,
                            "m_Z projection": proj,
                            "c": None if proj == "ols" else c,
                            "n_selected_Z": int(res.first_stage.supports["Z"].sum()),
                            "F": res.stat,
                            "boot_cv95": res.critical_value,
                            "boot_p": res.pvalue,
                            "alpha": res.alpha,
                            "theta": res.theta,
                            "theta_se": res.theta_se,
                        }
                    )
                except Exception as exc:  # pragma: no cover - degenerate cell
                    rows.append(
                        {"lags": lags, "m_Z projection": proj, "c": c, "error": str(exc)[:60]}
                    )
                if proj == "ols":
                    break  # the penalty is irrelevant without penalisation
    out = pd.DataFrame(rows)
    if "theta" in out:
        finite = out["theta"].dropna()
        out.attrs["theta_sign_flips"] = bool(
            len(finite) and (finite.min() < 0 < finite.max())
        )
    return out

File: r/ardldml/R/diagnostics.R
#' The Trend-Absorption Diagnostic
#'
#' The most important thing in the package to actually run, and the easiest to
#' skip.
#'
#' Residualisation is valid only when the controls remove \emph{nuisance}
#' stochastic trends, not the equilibrium relation being tested. If a control is
#' itself part of the equilibrium system, partialling it out absorbs the very
#' trend the test is meant to detect, and a non-rejection then reflects
#' over-absorption rather than the absence of a long-run relationship. That
#' requirement cannot be verified directly; this contrast is the practical
#' substitute.
#'
#' Four fits are run: the full and reduced control sets crossed with the
#' adaptive and the unpenalised level projection. Writing \eqn{p_{ad}} and
#' \eqn{p_{ols}} for the bootstrap p-values under the two projections and
#' \eqn{p_{full}} and \eqn{p_{red}} for those under the two control sets, the
#' diagnostic is the pair of gaps \eqn{\Delta_m = p_{ols} - p_{ad}} and
#' \eqn{\Delta_W = p_{full} - p_{red}}, read together with the stability of the
#' long-run coefficient.
#'
#' A large positive \eqn{\Delta_W} -- a verdict that rejects under the reduced
#' set but not under the full set -- together with a long-run coefficient that is
#' more sharply estimated under the reduced fit, is evidence that the nuisance
#' space is absorbing part of the tested relation, and the reduced-set verdict is
#' then the more credible one. Concordant verdicts across the four fits indicate
#' the conclusion is not an artefact of over-absorption.
#'
#' This is a hypothesis-generating device, not a formal test. It has no size and
#' no power; it tells you where to look.
#'
#' @param y Numeric vector: outcome in levels.
#' @param d Numeric vector: focal regressor in levels.
#' @param W Matrix or data frame of controls in levels.
#' @param drop Character vector of controls to omit from the reduced set. Choose
#'   these on \emph{economic} grounds: the ones most likely to be cointegrated
#'   with the tested relation itself.
#' @param B Bootstrap replications per fit. Four bootstraps are run, so the cost
#'   is roughly \code{4 * B} statistic evaluations.
#' @param seed Base seed; each fit is offset so the four bootstraps are
#'   independent.
#' @param progress Report which fit is running.
#' @param ... Passed to \code{\link{dml_spec}}; \code{adaptive} is set per fit.
#'
#' @return An object of class \code{"trend_absorption"}.
#'
#' @examples
#' \donttest{
#' df <- passthrough_regime("1999-2007")
#' ta <- trend_absorption(df$cpi, df$neer, as.matrix(df[, CONTROLS]),
#'                        drop = REDUCED_DROP, B = 19, seed = 1,
#'                        lags = 4, n_blocks = 5, buffer = 6,
#'                        integrated = DEFAULT_INTEGRATED)
#' ta
#' }
#'
#' @references
#' Villena, M. J. (2026). Testing cointegration with many persistent controls.
#' SSRN working paper. \doi{10.2139/ssrn.6472826}
#' @export
trend_absorption <- function(y, d, W, drop, B = 999L, seed = NULL,
                             progress = FALSE, ...) {
  W <- as.matrix(W)
  drop <- intersect(drop, colnames(W))
  W_red <- W[, setdiff(colnames(W), drop), drop = FALSE]

  args <- list(...)
  integrated <- args$integrated
  args$integrated <- NULL
  integrated_red <- if (is.null(integrated)) NULL else setdiff(integrated, drop)

  plan <- list(
    list(key = "full_adaptive", W = W, adaptive = TRUE, integ = integrated, off = 0L),
    list(key = "full_ols", W = W, adaptive = FALSE, integ = integrated, off = 1L),
    list(key = "reduced_adaptive", W = W_red, adaptive = TRUE, integ = integrated_red, off = 2L),
    list(key = "reduced_ols", W = W_red, adaptive = FALSE, integ = integrated_red, off = 3L)
  )

  fits <- list()
  for (p in plan) {
    if (progress) message("  fitting ", p$key, " (", ncol(p$W), " controls)...")
    call_args <- c(list(y = y, d = d, W = p$W, B = B, seed = if (is.null(seed)) NULL else seed + 10000L * p$off),
                   args, list(adaptive = p$adaptive, integrated = p$integ))
    fits[[p$key]] <- do.call(dml_bounds, call_args)
  }

  structure(list(
    fits = fits,
    dropped = drop,
    delta_m = fits$full_ols$boot$pvalue - fits$full_adaptive$boot$pvalue,
    delta_W = fits$full_adaptive$boot$pvalue - fits$reduced_adaptive$boot$pvalue
  ), class = "trend_absorption")
}

#' @rdname trend_absorption
#' @param x A \code{"trend_absorption"} object.
#' @export
as.data.frame.trend_absorption <- function(x, ...) {
  do.call(rbind, lapply(names(x$fits), function(k) {
    r <- x$fits[[k]]
    parts <- strsplit(k, "_", fixed = TRUE)[[1L]]
    data.frame(controls = parts[1L], projection = parts[2L],
               n_controls = r$n_controls, F = r$stat,
               boot_cv95 = if (is.null(r$boot)) NA_real_ else r$boot$crit,
               boot_p = if (is.null(r$boot)) NA_real_ else r$boot$pvalue,
               verdict = decision(r), alpha = r$alpha,
               theta = r$theta, theta_se = r$theta_se,
               stringsAsFactors = FALSE)
  }))
}

#' @rdname trend_absorption
#' @param level Significance level.
#' @export
absorption_verdict <- function(x, level = 0.05) {
  full <- x$fits$full_adaptive
  red <- x$fits$reduced_adaptive
  if (is.null(full$boot) || is.null(red$boot)) return("no bootstrap run")
  rej_full <- full$boot$pvalue < level
  rej_red <- red$boot$pvalue < level

  if (rej_red && !rej_full && x$delta_W > 0) {
    msg <- sprintf(paste0("Possible over-absorption. The reduced set rejects (p = %.3f) ",
                          "where the full set does not (p = %.3f); Delta_W = %+.3f."),
                   red$boot$pvalue, full$boot$pvalue, x$delta_W)
    if (red$theta_se < full$theta_se) {
      msg <- paste0(msg, sprintf(paste0(" The long-run coefficient is also more sharply ",
                                        "estimated under the reduced set (se %.3f versus %.3f), ",
                                        "which strengthens the reading. The reduced-set verdict ",
                                        "is the more credible."), red$theta_se, full$theta_se))
    } else {
      msg <- paste0(msg, " The long-run coefficient is not more sharply estimated under ",
                    "the reduced set, so ordinary specification sensitivity cannot be ruled out.")
    }
    return(msg)
  }
  if (rej_red == rej_full) {
    return(sprintf(paste0("Concordant verdicts across control sets (full p = %.3f, ",
                          "reduced p = %.3f). The conclusion does not appear to be an ",
                          "artefact of over-absorption."),
                   full$boot$pvalue, red$boot$pvalue))
  }
  sprintf(paste0("Discordant, but not in the over-absorption direction (full p = %.3f, ",
                 "reduced p = %.3f). Treat both verdicts as fragile."),
          full$boot$pvalue, red$boot$pvalue)
}

#' @rdname trend_absorption
#' @export
print.trend_absorption <- function(x, ...) {
  cat("Trend-absorption diagnostic\n")
  cat("dropped from reduced set:",
      if (length(x$dropped)) paste(x$dropped, collapse = ", ") else "(none)", "\n\n")
  print(as.data.frame(x), row.names = FALSE, digits = 4)
  th <- vapply(x$fits, function(f) f$theta, numeric(1))
  th <- th[is.finite(th)]
  cat(sprintf("\nDelta_m = p_ols  - p_ad  = %+.4f\n", x$delta_m))
  cat(sprintf("Delta_W = p_full - p_red = %+.4f\n", x$delta_W))
  if (length(th)) cat(sprintf("theta spread across fits = %.4f\n", max(th) - min(th)))
  cat("\n", absorption_verdict(x), "\n", sep = "")
  invisible(x)
}

#' Sweep the Penalty and the Projection
#'
#' The paper's robustness tables vary the penalty, the lag order and the level
#' projection, and a verdict can turn on that choice. Reporting a single cell
#' from this grid is specification search; reporting the grid is the method.
#'
#' The column to watch is \code{n_selected_Z}, the number of control
#' \strong{levels} the level projection retained. It is the empirical
#' counterpart of the effective integrated count: zero means nothing was
#' absorbed, so the test sits at the classical corner and orthogonalisation did
#' nothing, while a large value means heavy absorption and the long-run
#' coefficient should be checked for instability. A long-run coefficient that
#' changes sign across the grid is a warning that the conditioning set, not the
#' data, is driving the answer.
#'
#' @inheritParams trend_absorption
#' @param rules Penalty rules to sweep: \code{"low"}, \code{"medium"},
#'   \code{"high"}.
#' @param lags_grid Short-run lag orders to try.
#' @param projections Estimators to compare: \code{"adaptive"} (the paper's),
#'   \code{"plain"} (vanilla \eqn{\ell_1}) and \code{"ols"} (the unpenalised
#'   benchmark).
#' @param B Bootstrap draws per cell, or \code{0} for statistics only, which is
#'   much faster.
#'
#' @return A data frame with one row per cell, carrying a
#'   \code{"theta_sign_flips"} attribute.
#'
#' @examples
#' \donttest{
#' df <- passthrough_regime("1999-2007")
#' penalty_sensitivity(df$cpi, df$neer, as.matrix(df[, CONTROLS]),
#'                     lags_grid = 4, n_blocks = 5, buffer = 6,
#'                     integrated = DEFAULT_INTEGRATED)
#' }
#' @export
penalty_sensitivity <- function(y, d, W, rules = c("low", "medium", "high"),
                                lags_grid = 4L,
                                projections = c("adaptive", "plain", "ols"),
                                B = 0L, seed = NULL, ...) {
  rows <- list()
  for (lg in lags_grid) {
    for (proj in projections) {
      for (setting in rules) {
        args <- list(...)
        args$lags <- lg
        if (proj == "ols") {
          args$penalised <- FALSE
        } else {
          args$penalised <- TRUE
          args$adaptive <- proj == "adaptive"
          args$penalty <- setting
        }
        res <- tryCatch(
          do.call(dml_bounds, c(list(y = y, d = d, W = W, B = B, seed = seed), args)),
          error = function(e) e
        )
        if (inherits(res, "error")) {
          rows[[length(rows) + 1L]] <- data.frame(
            lags = lg, projection = proj, penalty = setting,
            n_selected_Z = NA_integer_, F = NA_real_, boot_p = NA_real_,
            alpha = NA_real_, theta = NA_real_, theta_se = NA_real_,
            stringsAsFactors = FALSE)
        } else {
          rows[[length(rows) + 1L]] <- data.frame(
            lags = lg, projection = proj,
            penalty = if (proj == "ols") "-" else setting,
            n_selected_Z = sum(res$supports$Z), F = res$stat,
            boot_p = if (is.null(res$boot)) NA_real_ else res$boot$pvalue,
            alpha = res$alpha, theta = res$theta, theta_se = res$theta_se,
            stringsAsFactors = FALSE)
        }
        if (proj == "ols") break   # the penalty is irrelevant without penalisation
      }
    }
  }
  out <- do.call(rbind, rows)
  fin <- out$theta[is.finite(out$theta)]
  attr(out, "theta_sign_flips") <- length(fin) > 0L && min(fin) < 0 && max(fin) > 0
  out
}
